rados/read_op_preview.go

```diff
@@ -0,0 +1,77 @@
+//go:build ceph_preview
+// +build ceph_preview
+
+package rados
+
+// #cgo LDFLAGS: -lrados
+// #include <rados/librados.h>
+// #include <stdlib.h>
+//
+import "C"
+
+import (
+	"unsafe"
+
+	"github.com/ceph/go-ceph/internal/cutil"
+)
+
+// ReadOpReadStep holds the result of the Read read operation.
+// Result is valid only after Operate() was called.
+type ReadOpReadStep struct {
+	// C returned data:
+	bytesRead *C.size_t
+	prval     *C.int
+
+	BytesRead int64 // Bytes read by this action.
+	Result    int   // Result of this action.
+}
+
+func (s *ReadOpReadStep) update() error {
+	s.BytesRead = (int64)(*s.bytesRead)
+	s.Result = (int)(*s.prval)
+
+	return nil
+}
+
+func (s *ReadOpReadStep) free() {
+	C.free(unsafe.Pointer(s.bytesRead))
+	C.free(unsafe.Pointer(s.prval))
+
+	s.bytesRead = nil
+	s.prval = nil
+}
+
+func newReadOpReadStep() *ReadOpReadStep {
+	return &ReadOpReadStep{
+		bytesRead: (*C.size_t)(C.malloc(cutil.SizeTSize)),
+		prval:     (*C.int)(C.malloc(cutil.IntSize)),
+	}
+}
+
+// Read bytes from offset into buffer.
+// `len(buffer)` is the maximum number of bytes read from the object.
+// `buffer[:ReadOpReadStep.BytesRead]` then contains object data.
+//  PREVIEW
+//
+// Implements:
+//  void rados_read_op_read(rados_read_op_t read_op,
+//                          uint64_t offset,
+//                          size_t len,
+//                          char * buffer,
+//                          size_t * bytes_read,
+//                          int * prval)
+func (r *ReadOp) Read(offset uint64, buffer []byte) *ReadOpReadStep {
+	oe := newReadStep(buffer, uint64(len(buffer)), offset)
```

Comment:
If the second argument is always len(buffer), we don't need it and can get `len(buffer)` in `newReadStep()`.

Reply:
Well, right now there are no other cases where the length differs. But maybe in the future there might be one, and all the occurrences would have to be refactored? I can't give examples where this would be true from the top of my head though...

Reply:
The Go idiomatic way would be to use a slice of the buffer then, like `buffer[:N]`, because that doesn't cause a copy. I think it would not be good coding style to give a function the whole buffer, but then tell it: "oh but please don't use the whole buffer but only the beginning of it".

And anyway I hope there will be not many occurrences of a function called `newReadStep()`. ;-)